Redirect some Nix commands output to a pager

thufschmitt opened the pull request:
Make `nix search`, `nix flake show` and `nix flake metadata` pipe their
output into a pager to make their result easier to visualise.

Fix #4548

edolstra commented:
So the reason I didn't enable paging yet is that it can interact badly with the progress indicator. Maybe what we should do is show the progress indicator until the first output is written, then disable the progress indicator and start the pager.

regnat commented:
> So the reason I didn't enable paging yet is that it can interact badly with the progress indicator. Maybe what we should do is show the progress indicator until the first output is written, then disable the progress indicator and start the pager.

I’ve actually disabled the progress indicator altogether when using the pager ([here](https://github.com/NixOS/nix/pull/4972/files#diff-aca69dfa28aaf6fff9e8629ac7240576bd8d5a3eab1c52ca30aea44d5e6da1bdR389)). But waiting for the first output might be nicer (I’ve no idea how to do that though :) ).

edolstra commented:
Ideally we would use `logger->writeToStdout` everywhere (instead of `std::cout`), and then the progress bar logger can disable itself and start the pager the first time `writeToStdout` is called.

regnat commented:
@edolstra I’ve rewritten the proxy mechanism in 8ad1ed5fb66b9443997aac761c1baf6771b30160 to wrap a logger rather than being its own different thing. That way we can nicely intercept `writeToStdout` to start the pager as needed and tell the underlying pager not to dump arbitrary stuff on stdeer anymore

SuperSandro2000 commented on src/nix/flake.cc.orig:
```diff
@@ -0,0 +1,1180 @@
+#include "command.hh"
```
You included the file by accident, right?

regnat replied:
Ooops, yes. Thanks, fixed

SuperSandro2000 commented:
Does less -R support the progress bar?

regnat commented:
> Does less -R support the progress bar?

`-R` wouldn’t. `-r` might, but it’s a bit hacky (and explicitly not recommended), so I’d rather not rely on that

Commit:
Make the pager a logger wrapper

Rewrite the pager mechanism to make it a (parametrized) instance of `Logger`
rather than being a standalone thing.
The main interest of this is that we have more control over when to
actually start it and make the inner pager more quiet (only when
`writeToStdout` is called)

Commit:
Set the log format to raw when the pager is used

The `bar` logger sents a bunch of things to stderr that messes-up the
pager output

Commit:
Redirect some Nix command outputs to a pager

Make `nix search`, `nix flake show` and `nix flake metadata` pipe their
output into a pager to make their result easier to visualise.

thufschmitt commented:
@edolstra friendly ping :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2023-08-04-nix-team-meeting-minutes-77/31487/1

edolstra commented:
Marking it as draft since it has a bunch of conflicts. Anybody want to pick this up?

tomberek commented:
stale: re-open and re-base as needed
